fastly/tls_mutual_authentication.go

```diff
@@ -0,0 +1,212 @@
+	PageSize int
+}
+
+// formatFilters converts user input into query parameters for filtering.
+func (i *ListTLSMutualAuthenticationsInput) formatFilters() map[string]string {
+	result := map[string]string{}
+	pairings := map[string]interface{}{
+		"include":      i.Include,
+		"page[size]":   i.PageSize,
+		"page[number]": i.PageNumber,
+	}
+
+	for key, value := range pairings {
+		switch t := reflect.TypeOf(value).String(); t {
+		case "string":
+			if value != "" {
+				result[key] = value.(string)
+			}
+		case "int":
+			if value != 0 {
+				result[key] = strconv.Itoa(value.(int))
+			}
+		}
+	}
+
+	return result
+}
+
+// ListTLSMutualAuthentication retrieves all resources.
+func (c *Client) ListTLSMutualAuthentication(i *ListTLSMutualAuthenticationsInput) ([]*TLSMutualAuthentication, error) {
+	p := "/tls/mutual_authentications"
+	filters := &RequestOptions{
+		Params: i.formatFilters(),
+		Headers: map[string]string{
+			"Accept": "application/vnd.api+json", // this is required otherwise the filters don't work
+		},
+	}
+
+	r, err := c.Get(p, filters)
+	if err != nil {
+		return nil, err
+	}
+
+	data, err := jsonapi.UnmarshalManyPayload(r.Body, reflect.TypeOf(new(TLSMutualAuthentication)))
+	if err != nil {
+		return nil, err
+	}
+
+	so := make([]*TLSMutualAuthentication, len(data))
+	for i := range data {
+		typed, ok := data[i].(*TLSMutualAuthentication)
+		if !ok {
+			return nil, fmt.Errorf("unexpected response type: %T", data[i])
+		}
+		so[i] = typed
+	}
+
+	return so, nil
+}
+
+// GetTLSMutualAuthenticationInput is used as input to the GetTLSMutualAuthentication function.
+type GetTLSMutualAuthenticationInput struct {
+	// ID is an alphanumeric string identifying a mutual authentication (required).
+	ID string
+	// Include is a comma-separated list of related objects to include (optional).
+	Include string
+}
+
+// GetTLSMutualAuthentication retrieves the specified resource.
+func (c *Client) GetTLSMutualAuthentication(i *GetTLSMutualAuthenticationInput) (*TLSMutualAuthentication, error) {
+	if i.ID == "" {
+		return nil, ErrMissingID
+	}
+
+	var ro *RequestOptions
+	if i.Include != "" {
+		ro = &RequestOptions{
+			Params: map[string]string{
+				"include": i.Include,
+			},
+		}
+	}
+
+	p := fmt.Sprintf("/tls/mutual_authentications/%s", i.ID)
+
+	r, err := c.Get(p, ro)
+	if err != nil {
+		return nil, err
+	}
+
+	var o TLSMutualAuthentication
+	if err := jsonapi.UnmarshalPayload(r.Body, &o); err != nil {
+		return nil, err
+	}
+
+	return &o, nil
+}
+
+// CreateTLSMutualAuthenticationInput is used as input to the CreateTLSMutualAuthentication function.
+type CreateTLSMutualAuthenticationInput struct {
+	// CertBundle is one or more certificates. Enter each individual certificate blob on a new line. Must be PEM-formatted (required).
+	CertBundle string `jsonapi:"attr,cert_bundle"`
+	// ID should not be set (it's internally used to help marshal the JSONAPI request data).
+	ID string `jsonapi:"primary,mutual_authentication"`
```

Comment:
I'm not familiar with JSON-API but is this a common convention?  Ideally we wouldn't be exposing internal fields like this in the API if we can avoid it.  One way to do this might be to make a private type inside `CreateTLSMutualAuthentication()` that embeds `CreateTLSMutualAuthenticationInput` and adds `ID`.

Something like:

```go
type CreateTLSMutualAuthenticationInput struct {
	CertBundle string `jsonapi:"attr,cert_bundle"`
	Name string `jsonapi:"attr,name,omitempty"`
}

func (c *Client) CreateTLSMutualAuthentication(i *CreateTLSMutualAuthenticationInput) (*TLSMutualAuthentication, error) {
	if i.CertBundle == "" {
		return nil, ErrMissingCertBundle
	}

	p := "/tls/mutual_authentications"

	in := struct{
		CreateTLSMutualAuthenticationInput
		ID `jsonapi:"primary,mutual_authentication"`
	}{
		CreateTLSMutualAuthenticationInput: i,
	}

	r, err := c.PostJSONAPI(p, in, nil)
	if err != nil {
		return nil, err
	}

	// ...
}
```

Reply:
> Ideally we wouldn't be exposing internal fields like this in the API if we can avoid it.

Agreed. 

I did this to keep it consistent with the rest of the code base which from what I can tell has always used this implementation.

I like the embed approach you've suggested as a way to internalise that ID 👍🏻 I'll give it a go locally and regenerate the fixtures and so if all works as expected it should just mean the dates in the regenerated fixtures change and nothing much else.

Reply:
@joeshaw This embedding approach doesn't appear to work btw...

```
tls_mutual_authentication_test.go:65: models should be a struct pointer or slice of struct pointers
```

Reply:
Just realised I wasn't passing a pointer `&in` so that moved the error onto...

```
    tls_mutual_authentication_test.go:65: 400 - Bad Request:

            Title:  Invalid value for cert_bundle
            Detail: can't be blank
```

Reply:
Spoke with Joe and he's OK with skipping this. If we could have made it work, then great but I tried 5 different ways to side-step the error stemming from the struct being embedded and ultimately it just wasn't compatible with how jsonapi marshalled the data.